trace-forward/src/Trace/Forward/Network/Acceptor.hs

```diff
@@ -0,0 +1,171 @@
+import           Ouroboros.Network.ErrorPolicy (nullErrorPolicies)
+import           Ouroboros.Network.IOManager (IOManager)
+import           Ouroboros.Network.Snocket (Snocket, localAddressFromPath, localSnocket)
+import           Ouroboros.Network.Socket (AcceptedConnectionsLimit (..),
+                                           SomeResponderApplication (..),
+                                           cleanNetworkMutableState, newNetworkMutableState,
+                                           nullNetworkServerTracers, withServerNode)
+import           Ouroboros.Network.Protocol.Handshake.Codec (cborTermVersionDataCodec,
+                                                             noTimeLimitsHandshake)
+import           Ouroboros.Network.Protocol.Handshake.Unversioned (UnversionedProtocol (..),
+                                                                   UnversionedProtocolData (..),
+                                                                   unversionedHandshakeCodec,
+                                                                   unversionedProtocolDataCodec)
+import           Ouroboros.Network.Protocol.Handshake.Type (Handshake)
+import           Ouroboros.Network.Protocol.Handshake.Version (acceptableVersion,
+                                                               simpleSingletonVersions)
+import           Ouroboros.Network.Util.ShowProxy (ShowProxy(..))
+
+import qualified Trace.Forward.Protocol.Acceptor as Acceptor
+import qualified Trace.Forward.Protocol.Codec as Acceptor
+import           Trace.Forward.Protocol.Type
+import           Trace.Forward.Queue (getTraceObjects)
+import           Trace.Forward.Configuration (AcceptorConfiguration (..), HowToConnect (..))
+
+listenToForwarder
+  :: (CBOR.Serialise lo,
+      ShowProxy lo,
+      Typeable lo)
+  =>  IOManager
+  -> AcceptorConfiguration lo
+  -> ([lo] -> IO ())
+  -> (NodeInfo -> IO ())
+  -> IO ()
+listenToForwarder iomgr config@AcceptorConfiguration{forwarderEndpoint} loHandler niHandler = do
+  let (LocalPipe localPipe) = forwarderEndpoint   
+      snocket = localSnocket iomgr localPipe
+      address = localAddressFromPath localPipe
+  doListenToForwarder snocket address noTimeLimitsHandshake app
+ where
+  app =
+    OuroborosApplication $ \_connectionId _shouldStopSTM ->
+      [ MiniProtocol
+          { miniProtocolNum    = MiniProtocolNum 1
+          , miniProtocolLimits = MiniProtocolLimits { maximumIngressQueue = maxBound }
+          , miniProtocolRun    = acceptTraceObjects config loHandler niHandler
+          }
+      ]
+
+doListenToForwarder
+  :: Ord addr
+  => Snocket IO fd addr
+  -> addr
+  -> ProtocolTimeLimits (Handshake UnversionedProtocol Term)
+  -> OuroborosApplication 'ResponderMode addr LBS.ByteString IO Void ()
+  -> IO ()
+doListenToForwarder snocket address timeLimits app = do
+  networkState <- newNetworkMutableState
+  nsAsync <- async $ cleanNetworkMutableState networkState
+  clAsync <- async . void $
```

Comment:
It will be cleaner (and safer) to use `race_` to run the two threads.

Reply:
Another remark, I don't think you actually need to run `withServerNode` in a separate thread. You could just use:
```
waitAnyCancel [nsAsync, serverAsync]
```
inside the call back to `withServerNode`.


If an async exception hits, you will leak both threads (that's why `race_` is safer).  What might want here is
```
withAsync (cleanNetworkMutableState networkState)
  $ \nsAsync ->
    withServerNode ... $ \serverAsync ->
      waitAny [nsAsync, serverAsync] -- no need for `waitAnyCancel`, `withAsync` and `withServerNode` will handle that

Reply:
Thanks, done.